analysis/Resonator_fitting/fit_biased_lorentzian.m

```diff
@@ -0,0 +1,186 @@
+parse(parser, freq, data, varargin{:});
+dims = parser.Results.dims;
+
+%validate inputs
+freq = freq(:);
+if isvector(data)
+    data = data(:);
+    N = length(data); M = 1;
+    if length(data) ~= length(freq)
+        error('Data and frequency dimensions do not match');
+    end
+elseif ~ismatrix(data)
+    error('Data has too many dimensions. I don''t know how to proceed.')
+else
+    [N, M] = size(data);
+    if N ~= length(freq) && M ~= length(freq)
+        error('Data dimensions do not match frequency vector dimensions.')
+    end
+    if strmatch('dims', parser.UsingDefaults)
+        if M == length(freq)
+            data = data';
+            t = M; M = N; N = t;
+        end
+    else
+        if dims == 2
+            data = data';
+            t = M; M = N; N = t;
+        end
+    end
+end
+
+options = optimset('MaxIter', parser.Results.MaxIter);
+
+fit_params = zeros(M, 5);
+fit_vars = zeros(M, 5);
+ssemin = zeros(M,1);
+
+for idx = 1:M  
+    p0 = initial_guess(freq, data(:,idx));
+    model = @(p)bias_lorentz_model(p, freq, data(:,idx));
+    [p, ssemin(idx)] = fminsearch(model, p0, options);
+    fit_params(idx, :) = p;
+    fit_vars(idx, :) = lorentz_variance(p, freq, ssemin(idx))';
+end
+
+if parser.Results.Plot 
+    
+    if M == 1
+        ffit = bias_lorentz(fit_params(1,:), freq);
+        [~, pband] = lorentz_variance(fit_params(1,:), freq, ssemin(1));
+        
+        figure();
+        hold on;
+        plot(freq, data, '.', 'MarkerSize', 15);
+        plot(freq, ffit, 'r-', 'LineWidth', 2);
+        plot(freq, ffit+1.96*pband, 'r--', 'LineWidth', 2);
+        plot(freq, ffit-1.96*pband, 'r--', 'LineWidth', 2);
+        grid on;
+        xlabel('Frequency', 'FontSize', 15);
+        ylabel('Amplitude', 'FontSize', 15);
+        title(sprintf('Biased Lorentzian Fit, f_0 = %.3f, FWHM = %.3f', fit_params(2), 2*fit_params(3)), 'FontSize', 12);
+        legend('Data', 'Fit', '95% C.I.');
+        set(gca, 'FontSize', 14);
+        
+    else
+        figure();
+        imagesc(freq, 1:M, data');
+        
+        fit = zeros(size(data));
+        for m=1:M
+            fit(:,m) = bias_lorentz(fit_params(m,:), freq);
+        end
+        
+        figure();
+        subplot(1,2,1);
+        imagesc(freq, 1:M, data'); axis square;
+        xlabel('Frequency', 'FontSize', 15);
+        title('Data', 'FontSize', 12);
+        set(gca, 'FontSize', 14);
+        subplot(1,2,2);
+        imagesc(freq, 1:M, fit'); axis square;
+        xlabel('Frequency', 'FontSize', 15);
+        title('Fit', 'FontSize', 12);
+        set(gca, 'FontSize', 14);
+        
+        figure();
+        subplot(2,1,1);
+        errorbar(1:M, fit_params(:,2), 1.96*sqrt(fit_vars(:,2)), '.', 'MarkerSize', 15, 'LineWidth', 2);
+        ylabel('f_0', 'FontSize', 15); set(gca, 'FontSize', 14); grid on;
+        subplot(2,1,2);
+        errorbar(1:M, fit_params(:,3), 1.96*sqrt(fit_vars(:,3)), '.', 'MarkerSize', 15, 'LineWidth', 2);
```

Comment:
here FWHM = params(3). Which one? I believe that it should be 2*params(3) if the data is in amplitudes